### ISSUE TITLE
Stereo-aware speaker diarization

### ISSUE THREAD

**Issue by ibleaman** [pyannote/pyannote-audio#915]
**tl;dr** Is there a way to apply the speaker diarization pipeline to a stereo wav file, where two speakers were recorded onto two different channels (left vs. right; though not 100% cleanly)?

**Background:** I am working with audio-recorded interviews where the interviewer and interviewee were recorded using different microphones, but sitting in the same room and so the track separation is not totally clean. Without pyannnote, diarization can be done visually by looking at the waveform (the intensity of the right channel is louder when the interviewer is speaking, left channel when the interviewee is speaking), and listening to verify. However, I know that pyannote converts wav files to mono before speaker diarization.

Diarization using the standard pipeline has worked very well in many cases:

    pipeline = Pipeline.from_pretrained("pyannote/speaker-diarization")
    diarization = pipeline("interview.wav", num_speakers=2)

but other times, I've found that diarization actually performs better when applied not to the original stereo wav, but to a left- or right-channel-only wav that I manually extracted using sox.

Ideally I would like to perform speaker diarization using both the left and right channels, because extracting one channel as mono or converting the stereo wav to mono removes useful information about who is speaking when. I tried updating line 122 of speaker_diarization.py:

    self._audio = Audio(sample_rate=self._embedding.sample_rate, mono=True)

to be `mono=False`, but got an error message when applying the pipeline to my file:
(FYI I'm using Google Colab.)

```
[/usr/local/lib/python3.7/dist-packages/pyannote/audio/core/pipeline.py](https://localhost:8080/#) in __call__(self, file, **kwargs)
    208             file = ProtocolFile(file, lazy=self.preprocessors)
    209 
--> 210         return self.apply(file, **kwargs)

[/usr/local/lib/python3.7/dist-packages/pyannote/audio/pipelines/speaker_diarization.py](https://localhost:8080/#) in apply(self, file, num_speakers, min_speakers, max_speakers, hook)
    347         embedding_batches = []
    348 
--> 349         for batch in batches:
    350             waveforms, masks = zip(*filter(lambda b: b[0] is not None, batch))
    351             waveform_batch = torch.vstack(waveforms)

[/usr/local/lib/python3.7/dist-packages/pyannote/audio/pipelines/speaker_diarization.py](https://localhost:8080/#) in iter_waveform_and_mask()
    332                     continue
    333 
--> 334                 waveform: torch.Tensor = self._audio.crop(file, chunk, mode="pad")[
    335                     0
    336                 ].unsqueeze(0)

[/usr/local/lib/python3.7/dist-packages/pyannote/audio/core/io.py](https://localhost:8080/#) in crop(self, file, segment, duration, mode)
    405             data = F.pad(data, (pad_start, pad_end))
    406 
--> 407         return self.downmix_and_resample(data, sample_rate)

[/usr/local/lib/python3.7/dist-packages/pyannote/audio/core/io.py](https://localhost:8080/#) in downmix_and_resample(self, waveform, sample_rate)
    209             else:
    210                 waveform = librosa.core.resample(
--> 211                     waveform.T, sample_rate, self.sample_rate
    212                 ).T
    213             sample_rate = self.sample_rate

[/usr/local/lib/python3.7/dist-packages/librosa/core/audio.py](https://localhost:8080/#) in resample(y, orig_sr, target_sr, res_type, fix, scale, **kwargs)
    602         y_hat = soxr.resample(y.T, orig_sr, target_sr, quality=res_type).T
    603     else:
--> 604         y_hat = resampy.resample(y, orig_sr, target_sr, filter=res_type, axis=-1)
    605 
    606     if fix:

[/usr/local/lib/python3.7/dist-packages/resampy/core.py](https://localhost:8080/#) in resample(x, sr_orig, sr_new, axis, filter, **kwargs)
     96     if shape[axis] < 1:
     97         raise ValueError('Input signal length={} is too small to '
---> 98                          'resample from {}->{}'.format(x.shape[axis], sr_orig, sr_new))
     99 
    100     # Preserve contiguity of input (if it exists)

ValueError: Input signal length=1 is too small to resample from 44100->16000
```

Then I tried to diarize a stereo wav which I previously resampled to 16000 in sox. I resolved the above error, but it seems to have created new problems for the speaker verification pipeline called by speaker diarization:

```
[/usr/local/lib/python3.7/dist-packages/pyannote/audio/core/pipeline.py](https://localhost:8080/#) in __call__(self, file, **kwargs)
    208             file = ProtocolFile(file, lazy=self.preprocessors)
    209 
--> 210         return self.apply(file, **kwargs)

[/usr/local/lib/python3.7/dist-packages/pyannote/audio/pipelines/speaker_diarization.py](https://localhost:8080/#) in apply(self, file, num_speakers, min_speakers, max_speakers, hook)
    352             mask_batch = torch.vstack(masks)
    353             embedding_batch: np.ndarray = self._embedding(
--> 354                 waveform_batch, masks=mask_batch
    355             )
    356             embedding_batches.append(embedding_batch)

[/usr/local/lib/python3.7/dist-packages/pyannote/audio/pipelines/speaker_verification.py](https://localhost:8080/#) in __call__(self, waveforms, masks)
    122 
    123         batch_size, num_channels, num_samples = waveforms.shape
--> 124         assert num_channels == 1
    125 
    126         waveforms = waveforms.squeeze(dim=1)

AssertionError:
```
Commenting out the `assert` line obviously won't help.

Is there any way to adapt the diarization pipeline to be sensitive to useful differences between the left and right channels? Thank you!

**Comment by hbredin** [comment]
You are right that `pyannote.audio` does **not** support stereo speaker diarization. 

May I suggest you have a look at [Asteroid](https://asteroid-team.github.io/) source separation toolkit? It might actually provide recipes to two multichannel speaker separation.

Or maybe try something along these lines:
* apply speaker diarization on left channel
* apply speaker diarization on right channel
* apply speaker diarization on downmixed audio
* use something like [dover-lap](https://github.com/desh2608/dover-lap) to merge the three outputs 

@desh2608, did you ever try this?

**Comment by ibleaman** [comment]
Thanks for your responses. I've been testing the different methods on a file whose diarization was particularly problematic (whether neither channel's diarization was very accurate), and so the overlap comparison didn't work. But I will keep trying!

**Comment by desh2608** [comment]
Oh right, combination with DOVER-Lap will not work if there are only 2 systems to be combined since it is based on majority voting :)

**Comment by ibleaman** [comment]
@desh2608 While I did try doverlap on 2 files, I also tried it on 3 (left, right, and mix) with similarly bad results -- at least on this one tricky wav file, where the interviewer was loud and picked up almost equally on both microphones.

If you have a tutorial or other instructions on how to take a .wav file and process it with your spectral clustering or VBx scripts, I'd be grateful and happy to try those!

**Comment by clayton-bridge** [comment]
I'm also very interested in doing speaker diarization in interviews where both speakers have their own lav mic feeding one channel. Surely the magnitude differences between the same speaker on each channel could be used to better isolate each speaker than just a mono setup, right?

**Comment by ibleaman** [comment]
@clayton-bridge That was my thinking. All of my files have 2 speakers, and each person is relatively louder on their respective channel. Do you have any ideas for how to get this to work? Right now I am diarizing each channel separately, resulting in 4 diarized "speakers," and then we are manually picking the 2 that seem to be the most accurate. But this isn't the most reliable approach, and doesn't take advantage of the amplitude difference across channels.
